drivers: i2c: Add Xilinx AXI I2C driver

robhancocksed opened the pull request:
Add a driver to support the Xilinx AXI IIC Bus Interface logic core, as described in Xilinx document PG090. This can be configured for use on most Xilinx FPGA-based platforms such as the Digilent Arty.

Both the 2.00a and 2.1 versions of the core are supported. The 2.00a revision has a bug causing dynamic read mode to not be reliable. With this version of the core the driver falls back to reading only 1 byte per interrupt.

robhancocksed committed:
dts: bindings: i2c: Added Xilinx I2C driver

Added device tree bindings for the Xilinx I2C driver (both 2.00a and 2.1
IP core revisions).

Signed-off-by: Robert Hancock <robert.hancock@calian.com>

robhancocksed committed:
drivers: i2c: add Xilinx AXI I2C driver

Add a driver to support the Xilinx AXI IIC Bus Interface logic core, as
described in Xilinx document PG090. This can be configured for use on
most Xilinx FPGA-based platforms such as the Digilent Arty.

Both the 2.00a and 2.1 versions of the core are supported. The 2.00a
revision has a bug causing dynamic read mode to not be reliable. With
this version of the core the driver falls back to reading only 1 byte
per interrupt.

Signed-off-by: Robert Hancock <robert.hancock@calian.com>

teburd commented on drivers/i2c/i2c_xilinx.c:
```diff
@@ -0,0 +1,641 @@
+	sys_write32(cr, config->base + REG_CR);
+	sys_write32((addr << 1) | TX_FIFO_START, config->base + REG_TX_FIFO);
+
+	/* TX FIFO empty detection is somewhat fragile because the status register
```
This I2C IP is a bit of a travesty in terms of not working in a nice way for a polling loop, but also not a nice way for interrupt driven transfers. Is this true for all versions of the IP?

robhancocksed replied:
AFAIK, yes. Both the TX FIFO empty detection and the RX FIFO full detection (in non-dynamic mode) have some annoying quirks. The fact that the driver can inadvertently set bits in the interrupt status register by writing 1 to them when they're not actually set is another annoyance I spent a bit of time figuring out. Most of these quirks are actually documented by Xilinx in some form at least..

This IP has been around quite a while at this point, so it seems unlikely to get a change for any of these if it might break driver compatibility. I think the only change they made in the 2.1 revision was to fix a bug where dynamic read mode sometimes didn't actually wait for the read byte count to be written to the FIFO before starting the transfer.

teburd commented on drivers/i2c/i2c_xilinx.c:
```diff
@@ -0,0 +1,641 @@
+
+	k_mutex_lock(&data->mutex, K_FOREVER);
+
+	ret = i2c_xilinx_wait_not_busy(config, data);
```
This is going to wait for an interrupt or 100ms, that seems like a pretty long time if there's no interrupt here? Unless I've mis-read. Aren't you already guaranteed that the IP block is idle at this point?

robhancocksed replied:
The IP block itself should be idle (unless perhaps there was a target-mode transfer in progress) but it's possible another controller using the bus when the transfer is started, so this waits for that to be complete. 100ms might be excessive, I basically picked that as an arbitrary value for "if it takes this long something is almost certainly broken" - open to another suggestion..

teburd reviewed (commented):
Without an associated board this driver won't be at a minimum built which would be nice to see this associated with. Without being built it's likely to bit-rot as API changes are made.

I get that this is probably used in the context of some soft-core setup with a microblaze maybe? Is it used anywhere else, with say the Zynq?

The code itself looks fine to me, the IP itself is... well I'm happy I'm not using it, looks like its full of "fun" quirks given the inline comments.

+1 from on seeing something build it.

robhancocksed reviewed (commented):
So this IP can be quite easily used with the Digilent Arty board which is in Zephyr, which uses the ARM Cortex-M1 or Cortex-M3 soft cores, and indeed that's been our development test platform for this. However, that board's device tree setup in the tree is presuming you are using the stock FPGA image from ARM for the DesignStart cores on the Arty, which doesn't instantiate the I2C core. So you'd have to modify that logic build to add the core into the block design in Vivado, rebuild the FPGA image and then add the core into the device tree at whatever address/interrupt you assigned it to.

Potentially the I2C core could be added in as another board variant or an overlay/shield or something? I'm not familiar enough with those to immediately know what might make sense.

The Digilent Zybo is another Zephyr board you could likely use this with fairly easily, but it looks like the Zephyr setup isn't using any programmable logic load at all (it's just set up to use the processor-side hard cores in the Zynq).

robhancocksed committed:
tests: drivers: build_all: add build for i2c_xilinx_axi

There is currently not an in-tree Zephyr board which enables the
i2c_xilinx_axi driver as part of its standard configuration. (It can be
enabled fairly easily with the Digilent Arty board, but that requires
modifying the ARM-provided FPGA block design in Vivado and rebuilding
the FPGA image, so it cannot be used with the default FPGA image.)

Add a build_all target to provide build coverage for this driver.

Signed-off-by: Robert Hancock <robert.hancock@calian.com>

robhancocksed commented on drivers/i2c/Kconfig.xilinx:
```diff
@@ -0,0 +1,11 @@
+	bool "Xilinx AXI I2C driver"
+	default y
+	depends on DT_HAS_XLNX_XPS_IIC_2_00_A_ENABLED || DT_HAS_XLNX_XPS_IIC_2_1_ENABLED
+	depends on EVENTS
```
Switched this to use select and got rid of CONFIG_EVENTS in the build_all.

ibirnbaum reviewed (changes requested):
I've just had a look at the PG090 docs and compared them to the Zynq's TRM (UG585). Just by looking at the register map, it's apparent that this IP core is not register compatible with the I2C controllers contained in both the Zynq-7000's and the ZynqMP's Processor System. Therefore, the driver file names, the Kconfig file name and the contained CONFIG_I2C_XILINX item should make it clear that this is Xilinx's AXI I2C, not Xilinx's PS I2C (for which a driver implementation exists within my company, gotta ask my colleague some time if it's ready for contribution...). Compare with the Xilinx GPIO drivers, where driver implementations exist for both the PS and the AXI GPIO.

ibirnbaum commented:
A few words on usefulness and testability: this wouldn't be the first driver for a Xilinx FPGA IP core, as mentioned, there's already the Xilinx AXI GPIO driver. In order to actually use this driver, an FPGA bitstream in which the corresponding FPGA IP is instantiated is required.

The Arty is an actual hardware target board which supports this, and I myself have been working on a board definition for the Zedboard, another popular Zynq-based board capable of instantiating Xilinx FPGA IP cores: https://github.com/ibirnbaum/zephyr/tree/zedboard.

This board definition so far hasn't been contributed yet, as, while I/O pin mapping can in the meantime be handled by a Zynq pinctrl driver, a clock control driver as suggested by @henrikbrixandersen is still WIP, at the time being, Therefore, the board and its peripherals aren't defined as cleanly as they could be (depending the least on certain pre-conditions regarding system initialization), and at the time being, the best possible way of managing all of the Zynq's peripherals can not yet be achieved - comp. #48243. Also, a driver for loading an FPGA bitstream containing the IP cores as part of the Zephyr startup doesn't exist yet.

This means that at the time being, in order to use a matching FPGA bitstream and to fully configure the Zynq peripherals' I/Os and clocks, a certain workflow is required, and the resulting binaries must be booted using either Xilinx's First-Stage Bootloader in conjunction with u-boot, or the FSBL may be substituted by the u-boot SPL. These components ensure that prior to booting Zephyr, both the Processor System and the Programmable Logic are fully set up. Under these circumstances, the AXI GPIO driver, and also the AXI I2C driver submitted here, will work. However, this entire process is closely linked to Xilinx's development tools. I've written a full tutorial for the Zedboard which shows how to provide a Processor System configuration for Zephyr, how to instantiate AXI GPIO IP Core instances, and how to build and boot the binaries for the Zedboard. The tutorial is available here: https://github.com/ibirnbaum/zephyr4zedboard-tutorial

Clear instructions regarding the required development tools and the required workflows are probably required whereever an FPGA IP core is to be used.

A driver for any of the Zynq's integrated Processor System peripherals can always be tested in loopback mode under qemu_cortex_a9 (simulation quality may vary), tests for FPGA ip cores aside from running on actual hardware under the afforementioned conditions will most likely always be dummy tests.

robhancocksed commented:
@ibirnbaum Good point, I've updated the config options, filenames and function names to add in the AXI reference.

robhancocksed commented:
I've added some more fixes to resolve some issues seen in our environment:

- Fix issues with write completion timeouts in some cases
- Always reinitialize the controller between transfers, like the Linux driver does
- Always wait for bus-not-busy indication after each STOP, rather than just after the transfers are complete

ibirnbaum commented:
> @ibirnbaum I think your requested changes have been addressed?

Almost - would you mind renaming the Kconfig file in drivers/i2c from Kconfig.xilinx to Kconfig.xlinx_axi so that we might add a Kconfig.xilinx_ps for the other I2C controller at some point and have a consistent naming scheme?

robhancocksed commented:
@ibirnbaum Updated as suggested now.

ibirnbaum commented:
> @ibirnbaum Updated as suggested now.

Thank you!
